conda/match.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from .models.dist import Dist
+from .models.index_record import IndexRecord
+from .common.compat import string_types, text_type, iteritems
+from .version import VersionSpec
+from .exceptions import CondaValueError
+from ast import literal_eval
+import re
+import sys
+
+class MatchSpec(object):
+
+    def __new__(cls, *specs, **kwargs):
+        if len(specs) == 1 and not kwargs and isinstance(specs[0], cls):
+            return specs[0]
+        normalize = kwargs.pop('normalize', False)
+        self = object.__new__(cls)
+        self.specs = {}
```

Comment:
What is this, and how is it used?  Seems pretty important.  I'd like to wrap it in `conda._vendor.auxlib.collection.frozendict` before we `return self` at the end of `__new__`.  Later, we can take advantage of frozendict being hashable.